docs/architecture/adr-012-ibc-relayer-v2.md

```diff
@@ -0,0 +1,744 @@
+assemble their own computer with the exact hardware specs that they prefer.
+However, even though having modular hardware components is very useful, the
+majority of consumers would prefer _not_ to assemble their own computer, or to
+understand how each hardware component works. Instead, consumers prefer to have
+pre-assembled computers that are designed to fit specific use cases, such as
+gaming laptops or smartphones with good cameras, and choose a model that matches
+closest to their use cases.
+
+To help normal users to build custom relayers with minimal effort, the relayer
+framework offers _all-in-one_ traits that have most parts of the relayer
+pre-configured as _presets_. The relayer framework currently offers two presets:
+minimal and full-featured.
+
+The minimal preset, as its name implies, offers an abstract implementation of a
+minimal relayer, which only performs the core logic of relaying with no
+additional complexity. Because the relayer is minimal, it has less requirements
+of what the concrete contexts need to implement. As a result, it takes the least
+effort for users to build a minimal relayer that targets custom chains like
+solomachines, or custom environments like Wasm.
+
+On the other hand, the full preset includes all available components that the
+relayer framework offers, such as message batching, parallel transactions,
+packet filter, error retry, and telemetry. While these features are not
+essential for a minimal relayer to work, they are useful for operating
+production relayers that require high reliability and efficiency. As a tradeoff,
+since the full relayer includes more components with complex logic, it also
+impose more requirements to the concrete context implementation, and there may
+be more potential for subtle bugs to be found.
+
+The minimal preset requires implementers of custom relay context to implement
+the _one-for-all_ traits such as `OfaBaseChain` and `OfaBaseRelay`. In addition
+to that, the full preset requires implementers to also implement traits like
+`OfaFullChain` and `OfaFullRelay`. An example use of this is demonstrated
+in the later section using the Cosmos chain context.
+
+### Limitations of All-In-One Traits
+
+Since the relayer framework only offers two presets, there is a gap between
+the minimal preset and the full preset. As a result, users cannot selectively
+choose only specific features from the full relayer, such as enabling only
+the message batching feature without telemetry. The choice of us imposing
+this restriction is _deliberate_, as we want to keep the all-in-one traits
+simple and have a smooth learning curve.
+
+It is worth noting that even though the relayer components defined using
+context-generic programming are fully modular, the all-in-one traits are
+designed to be _rigid_ and _specialized_. Observant readers may notice that
+the all-in-one traits are similar to the traditional god traits that are
```

Comment:
"Traditional god traits" seems like a typo, but I have no idea what's it supposed to be instead.

Reply:
They are just the traits that contain too many methods, with developers keep adding new methods to it whenever they need something new. I have changed the term to monolithic traits and explain them a little bit.